pkg/skaffold/deploy/kubectl.go

```diff
@@ -82,123 +83,179 @@ func NewKubectlDeployer(cfg *v1alpha2.DeployConfig, kubeContext string) *Kubectl
 // Deploy templates the provided manifests with a simple `find and replace` and
 // runs `kubectl apply` on those manifests
 func (k *KubectlDeployer) Deploy(ctx context.Context, out io.Writer, b *build.BuildResult) (*Result, error) {
-	if len(k.DeployConfig.KubectlDeploy.Manifests) == 0 {
-		if len(b.Builds) != 1 {
-			return nil, errors.New("must specify manifest if using more than one image")
-		}
-		yaml, err := generateManifest(b.Builds[0])
-		if err != nil {
-			return nil, errors.Wrap(err, "generating manifest")
-		}
-
-		if err := k.deployManifestFile(strings.NewReader(yaml), []build.Build{{ImageName: "IMAGE", Tag: b.Builds[0].Tag}}); err != nil {
-			return nil, errors.Wrap(err, "deploying manifest")
-		}
-		return &Result{}, nil
+	manifests, err := k.readOrGenerateManifests(b)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading manifests")
 	}
-	manifests, err := util.ExpandPathsGlob(k.DeployConfig.KubectlDeploy.Manifests)
+
+	manifests, err = manifests.replaceImages(b.Builds)
 	if err != nil {
-		return nil, errors.Wrap(err, "expanding kubectl manifest paths")
+		return nil, errors.Wrap(err, "replacing images in manifests")
 	}
-	for _, m := range manifests {
-		logrus.Debugf("Deploying path: %s", m)
-		if err := k.deployManifest(out, b.Builds, m); err != nil {
-			return nil, errors.Wrap(err, "deploying manifests")
-		}
+
+	err = k.kubectl(manifests.reader(), out, "apply", "-f", "-")
+	if err != nil {
+		return nil, errors.Wrap(err, "deploying manifests")
 	}
 
 	return &Result{}, nil
 }
 
-func generateManifest(b build.Build) (string, error) {
-	logrus.Info("No manifests specified. Generating a deployment.")
-	dockerfilePath := filepath.Join(b.Artifact.Workspace, b.Artifact.DockerArtifact.DockerfilePath)
-	r, err := os.Open(dockerfilePath)
-	if err != nil {
-		return "", errors.Wrap(err, "reading dockerfile")
+// Cleanup deletes what was deployed by calling Deploy.
+func (k *KubectlDeployer) Cleanup(ctx context.Context, out io.Writer) error {
+	if len(k.KubectlDeploy.Manifests) == 0 {
+		return k.kubectl(nil, out, "delete", "deployment", "skaffold")
```

Comment:
if you end up using `generateName` here, we'll have to delete by label or annotation.